x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java

```diff
@@ -200,14 +200,34 @@ public void addListener(ActionListener<Long> listener, long value) {
         assert invariant();
     }
 
-    private static void executeListener(final ActionListener<Long> listener, final Supplier<Long> result) {
+    /**
+     * Return the result of this future, if it has been completed successfully, or unwrap and throw the exception with which it was
+     * completed exceptionally. It is not valid to call this method if the future is incomplete.
+     */
+    private Long actionResult() throws Exception {
+        try {
+            return result();
+        } catch (ExecutionException e) {
+            if (e.getCause() instanceof Exception exCause) {
+                throw exCause;
+            } else {
+                throw e;
+            }
+        }
+    }
```

Comment:
I moved this down the hierarchy as it helps avoid excessive nesting on execution exceptions here